querydsl instanceof results in wrong query for Type(?1)

copied from SE (http://stackoverflow.com/questions/27941534/)

If I use Hibernate entity Inheritance feature, I could easily always try to load only the subClass I need like this

    SELECT p FROM Product p WHERE TYPE(p) = ?1

trying to achieve the same query with queryDSL:

    QProduct.product.instanceOfAny(clazz)

But I get the following exception

    Caused by: java.lang.IllegalArgumentException: Parameter value [Product] did not match expected type [java.lang.Integer (n/a)]

query that querydsl produces is

    SELECT count(product)
    FROM Product product
    WHERE product.class = ?1  // this should be TYPE(product) = ?1


my enitity

    @Entity
    @Table(name = "product")
    @Inheritance(strategy=InheritanceType.JOINED)
    public abstract class Product {
        
    	...

     

Full Stack Trace (for a similar case but different Class name):

    Caused by: java.lang.IllegalArgumentException: Parameter value [BasketDiscount] did not match expected type [java.lang.Integer (n/a)]
    	at org.hibernate.jpa.spi.BaseQueryImpl.validateBinding(BaseQueryImpl.java:874)
    	at org.hibernate.jpa.internal.QueryImpl.access$000(QueryImpl.java:80)
    	at org.hibernate.jpa.internal.QueryImpl$ParameterRegistrationImpl.bindValue(QueryImpl.java:248)
    	at org.hibernate.jpa.internal.QueryImpl$JpaPositionalParameterRegistrationImpl.bindValue(QueryImpl.java:337)
    	at org.hibernate.jpa.spi.BaseQueryImpl.setParameter(BaseQueryImpl.java:674)
    	at org.hibernate.jpa.spi.AbstractQueryImpl.setParameter(AbstractQueryImpl.java:198)
    	at org.hibernate.jpa.spi.AbstractQueryImpl.setParameter(AbstractQueryImpl.java:49)
    	at com.mysema.query.jpa.impl.JPAUtil.setConstants(JPAUtil.java:55)
    	at com.mysema.query.jpa.impl.AbstractJPAQuery.createQuery(AbstractJPAQuery.java:128)
    	at com.mysema.query.jpa.impl.AbstractJPAQuery.count(AbstractJPAQuery.java:79)
    	at org.springframework.data.jpa.repository.support.QueryDslJpaRepository.findAll(QueryDslJpaRepository.java:125)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:606)
    	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.executeMethodOn(RepositoryFactorySupport.java:442)
    	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:427)
    	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:381)
    	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
    	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:98)
    	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:262)
    	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:95)
    	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
    	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:136)
    	... 94 more

--- a/querydsl-jpa/src/main/java/com/querydsl/jpa/HQLTemplates.java
+++ b/querydsl-jpa/src/main/java/com/querydsl/jpa/HQLTemplates.java
@@ -75,10 +75,6 @@ public class HQLTemplates extends JPQLTemplates {
         builder.put(BigDecimal.class, "big_decimal");
         typeNames = builder.build();
 
-        // TODO : remove this when Hibernate supports type(alias)
-        add(Ops.INSTANCE_OF, "{0}.class = {1}");
-        // TODO : remove this when Hibernate supports type(alias)
-        add(JPQLOps.TYPE, "{0}.class");
         // TODO : remove this when Hibernate supports member of properly
         add(JPQLOps.MEMBER_OF, "{0} in elements({1})");
         add(JPQLOps.NOT_MEMBER_OF, "{0} not in elements({1})");
@@ -108,11 +104,6 @@ public class HQLTemplates extends JPQLTemplates {
     @Override
     public boolean wrapElements(Operator<?> operator) {
         return wrapElements.contains(operator);
-    }
-
-    @Override
-    public boolean isTypeAsString() {
-        return true;
     }
 
     @Override

--- a/querydsl-jpa/src/main/java/com/querydsl/jpa/JPQLSerializer.java
+++ b/querydsl-jpa/src/main/java/com/querydsl/jpa/JPQLSerializer.java
@@ -417,9 +417,6 @@ public class JPQLSerializer extends SerializerBase<JPQLSerializer> {
                 super.visitOperation(type, operator, args);
             }
 
-        } else if (operator == Ops.INSTANCE_OF) {
-            visitInstanceOf(type, operator, args);
-
         } else if (operator == Ops.NUMCAST) {
             visitNumCast(args);
 
@@ -454,23 +451,6 @@ public class JPQLSerializer extends SerializerBase<JPQLSerializer> {
         final Class<?> targetType = (Class<?>) ((Constant<?>) args.get(1)).getConstant();
         final String typeName = templates.getTypeForCast(targetType);
         visitOperation(targetType, JPQLOps.CAST, ImmutableList.of(args.get(0), ConstantImpl.create(typeName)));
-    }
-
-    private void visitInstanceOf(Class<?> type, Operator<?> operator,
-            List<? extends Expression<?>> args) {
-        if (templates.isTypeAsString()) {
-            final List<Expression<?>> newArgs = new ArrayList<Expression<?>>(args);
-            final Class<?> cl = ((Class<?>) ((Constant<?>) newArgs.get(1)).getConstant());
-            // use discriminator value instead of fqnm
-            if (cl.isAnnotationPresent(DiscriminatorValue.class)) {
-                newArgs.set(1, ConstantImpl.create(cl.getAnnotation(DiscriminatorValue.class).value()));
-            } else {
-                newArgs.set(1, ConstantImpl.create(cl.getSimpleName()));
-            }
-            super.visitOperation(type, operator, newArgs);
-        } else {
-            super.visitOperation(type, operator, args);
-        }
     }
 
     private void visitPathInCollection(Class<?> type, Operator<?> operator,

--- a/querydsl-jpa/src/main/java/com/querydsl/jpa/JPQLTemplates.java
+++ b/querydsl-jpa/src/main/java/com/querydsl/jpa/JPQLTemplates.java
@@ -146,11 +146,6 @@ public class JPQLTemplates extends Templates {
         return false;
     }
 
-    public boolean isTypeAsString() {
-        // TODO : get rid of this when Hibernate supports type(alias)
-        return false;
-    }
-
     public String getTypeForCast(Class<?> cl) {
         return cl.getSimpleName().toLowerCase();
     }

